Mathlib/RingTheory/Congruence/Hom.lean

```diff
@@ -0,0 +1,485 @@
+  refine ⟨fun H ↦ ext'' (Setoid.ker_eq_lift_of_injective f h H).symm, ?_⟩
+  rintro H ⟨x⟩ ⟨y⟩
+  simp [H]
+
+theorem lift_bijective_iff {h : c ≤ ker f} :
+    Function.Bijective (c.lift f h) ↔ c = ker f ∧ Surjective f := by
+  unfold Function.Bijective
+  simp only [lift_injective_iff, lift_surjective_iff]
+
+/-- Given a ring homomorphism `f` from `M` to `P`, the kernel of `f` is the
+unique ring congruence relation on `M` whose induced map from the quotient of
+`M` to `P` is injective. -/
+theorem ker_eq_lift_of_injective (H : c ≤ ker f) (h : Injective (c.lift f H)) : ker f = c :=
+  (lift_injective_iff.mp h).symm
+
+variable (f)
+
+/-- The homomorphism induced on the quotient of a ring by the kernel of a ring homomorphism. -/
+def kerLift : (ker f).Quotient →+* P :=
+  (ker f).lift f fun _ _ => id
+
+variable {f}
+
+/-- The diagram described by the universal property for quotients of rings, when
+the ring congruence relation is the kernel of the homomorphism, commutes. -/
+theorem kerLift_mk (x : M) : kerLift f x = f x :=
+  rfl
+
+/-- A ring homomorphism `f` induces an injective homomorphism on the quotient by `f`'s kernel. -/
+theorem kerLift_injective (f : M →+* P) : Injective (kerLift f) :=
+  AddCon.kerLift_injective (f : M →+ P)
+
+/-- Given ring congruence relations `c, d` on a ring such that `d` contains `c`,
+`d`'s quotient map induces a homomorphism from the quotient by `c` to the
+quotient by `d`. -/
+def map (c d : RingCon M) (h : c ≤ d) : c.Quotient →+* d.Quotient :=
+  c.lift d.mk' fun x y hc => show ker d.mk' x y from (ker_mk'_eq d).symm ▸ h hc
+
+/-- Given ring congruence relations `c, d` on a ring such that `d` contains `c`,
+the definition of the homomorphism from the quotient by `c` to the quotient by
+`d` induced by `d`'s quotient map. -/
+theorem map_apply {c d : RingCon M} (h : c ≤ d) (x) :
+    c.map d h x = c.lift d.mk' (fun _ _ hc => d.eq.2 <| h hc) x :=
+  rfl
+
+end
+
+section
+
+variable [NonAssocSemiring M] [NonAssocSemiring N] [NonAssocSemiring P]
+
+variable {c : RingCon M}
+
+theorem rangeS_mk' : RingHom.rangeS c.mk' = ⊤ :=
+  RingHom.rangeS_eq_top.mpr (mk'_surjective _)
+
+variable {f : M →+* P}
+
+/-- Given a congruence relation `c` on a semiring and a homomorphism
+`f` constant on the equivalence classes of `c`, `f` has the same image
+as the homomorphism that `f` induces on the quotient. -/
+theorem rangeS_lift (H : c ≤ ker f) :
+    RingHom.rangeS (c.lift f H) = f.rangeS :=
+  SetLike.coe_injective <| Set.range_quot_lift _
+
+/-- Given a ring homomorphism `f`, the induced homomorphism
+on the quotient by `f`'s kernel has the same image as `f`. -/
+theorem kerLift_rangeS_eq : RingHom.rangeS (kerLift f) = RingHom.rangeS f :=
+  rangeS_lift fun _ _ => id
+
+variable (c)
+
+/-- The **first isomorphism theorem for semirings**. -/
+noncomputable def quotientKerEquivRangeS (f : M →+* P) :
+    (ker f).Quotient ≃+* f.rangeS where
+  __ := RingHom.codRestrict (kerLift f) _ _
+  __ := Setoid.quotientKerEquivRange _
```

Comment:
Can you add `↑(quotientKerEquivRangeS f (↑x)) = f x` (and also for the other defs below)? Thanks.